### ISSUE TITLE
COLUMN ORDERING

### ISSUE THREAD

**Issue by adrianON** [metabase/metabase#4329]
Hello everyone, 
When building a question the columns are ordered alphabetically (A,B,C etc), I also do not see any ability in the Admin panel to modify column order.  I see where this functionality was added to later builds of metabase but I do not see it in version in 0.22.2 using Postgres.  What am I missing?  Here are the particulars:

**BROWSER**: Version 55.0.2883.75 m (64-bit),
**OPERATING** SYSTEM: WINDOWS 10 also tried on CENTOS 6
**DATABASE**:  POSTGRES 9.4
**METABASE VERSION**:  0.22.2
**METABASE HOSTING ENVIRONMENT**: WINDOWS 10 and tried CENTOS 6
**METABASE INTERNAL DATABAS**E - DEFAULT H2

**Comment by camsaul** [comment]
@adrianON you can change the order of columns once you run a query by clicking Chart Settings option (the little gear in the top-left of the chart). Read more about chart settings [here](http://www.metabase.com/docs/LATEST/users-guide/04-visualizing-results.html).

**Comment by adrianON** [comment]
Thanks for your answer.  I was looking right at it yesterday prior to asking the question.  I missed that you had to drag and drop to change the column order.  Metabase has excellent functionality and very easy to setup and build dashboards and reports.

**Comment by camsaul** [comment]
I agree 😉

Closing since it sounds like this question has been answered

**Comment by anothergituser** [comment]
Well how about an option for original table order and also original table fields names. If i spend a lot of time watching my tables in mysql for example i kinda know where to find each field, and i don't like the A-Z order, also when having lots of fields means the gear icon config where i can drag the fields to order is cumbersome and i would need to do it for each new query. Also i would appreciate having the column names as they were defined. Thanks

**Comment by anki-code** [comment]
> i don't like the A-Z order
+1

**Comment by lindsay-stevens** [comment]
@adrianON Same issue for me, Postgres 9.6 data source, Metabase v0.22 with Postgres 9.6 backend.

There seems to be 3 ways that field ordering happens, depending on where you're looking:

1. in the Admin Panel -> Data Model:
  - sorted by position, then ABC (see src/metabase/query_processor/annotate.clj - "special_type" seems to have no effect on order)
  - can't be updated in the UI (despite the comment in the above annotate.clj)
  - can be overridden in the database by setting the "position" integer (metabase_field.position)
  - position defaults to 0 (zero), is not set during introspection, is not exposed in Field API or in UI
  - can set position in the Metabase db with UPDATEs
2. in the Data Reference:
  - sorted by ID (metabase_field.id)
  - can't be changed in the UI
  - can re-order the field IDs in the Metabase db with UPDATEs (swapping IDs via 0)
3. in queries, 
  - sorted by position, then ABC (same method as Admin Panel -> Data Model)
  - can be overridden for a particular query in the UI (or a PUT to the card API), to save the column order in a JSON array under (report_card.visualization_settings), e.g.

```json
{"table.column_widths":[],
 "table.columns":[
    {"name":"my_col","enabled":true},
    {"name":"my_other_col","enabled":true},
    {"name":"col_three","enabled":true}]}
```

My use case is that I've got a database with a bunch of queries, and Metabase is one of many ways in which that data is consumed. The column order is important to making sense of the queries, especially when it is a rollup / cube type of query where the row header / grouping columns are first. The data reference is a useful feature here, to have an integrated documentation of the queries in Metabase.

@camsaul - I'm wondering if there's an intention to consolidate the default field ordering, perhaps around the "position" field (with the ability to customise in cards), or if "position" is a remnant of something that was removed?

**Comment by camsaul** [comment]
@lindsay-stevens an older version of Metabase let you set the default ordering of columns in a table which is the purpose of the `position` column you've seen. I don't remember when (or why) that was taken out (@metabase/core-developers?).

**Comment by GillianHanson** [comment]
+1 to the above question. Is there a way to retain column order or specify column order in the current version?

**Comment by MrZeroo00** [comment]
+1, I'm having the same issue.

**Comment by arcreative** [comment]
Can we please have this addressed/reopened?  There's literally no logic in not letting users retain the original sorting of the table.  Quite frankly, I use Metabase for convenience and to save time, and taking 5 minutes to re-order every single table to match _*something you already know*_ is completely ridiculous.  There's literally no logic that a `start_date` shouldn't appear before `end_date` in my table of 50 columns, and having to do this risks completely borking the interpretation of the data, and thus the entire purpose of BI.

**Comment by arcreative** [comment]
Looks like this is being addressed in #3720.
